File: sdk-python/rtmemory/graph.py
"""GraphNamespace — async methods for the /v1/graph/ API."""

from __future__ import annotations

from typing import Optional

import httpx

from rtmemory.types import GraphNeighborhood


class GraphNamespace:
    """Namespace for graph visualization and traversal."""

    def __init__(self, http: httpx.AsyncClient) -> None:
        self._http = http

    async def neighborhood(
        self,
        entity_id: str,
        space_id: Optional[str] = None,
        max_hops: int = 3,
        relation_types: Optional[list[str]] = None,
        direction: str = "both",
    ) -> GraphNeighborhood:
        """Get the neighborhood subgraph around an entity.

        Calls GET /v1/graph/neighborhood with the specified parameters.
        """
        params: dict[str, str | int] = {
            "entity_id": entity_id,
            "max_hops": max_hops,
            "direction": direction,
        }
        if space_id is not None:
            params["space_id"] = space_id
        if relation_types is not None:
            params["relation_types"] = ",".join(relation_types)
        resp = await self._http.get("/v1/graph/neighborhood", params=params)
        resp.raise_for_status()
        return GraphNeighborhood.model_validate(resp.json())

File: sdk-js/src/graph.ts
/**
 * GraphClient — thin wrapper around graph visualization API endpoints.
 */

import type { GraphNeighborhood } from "./types";
import { RTMemoryError } from "./client";

export class GraphClient {
  private baseUrl: string;
  private headers: Record<string, string>;

  constructor(baseUrl: string, headers: Record<string, string>) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.headers = { ...headers };
  }

  /** Get the neighborhood of an entity in the knowledge graph. */
  async neighborhood(params: {
    entityId: string;
    spaceId?: string;
    maxHops?: number;
    relationTypes?: string[];
    direction?: string;
  }): Promise<GraphNeighborhood> {
    const url = new URL(`${this.baseUrl}/v1/graph/neighborhood`);
    url.searchParams.set("entity_id", params.entityId);
    url.searchParams.set("max_hops", String(params.maxHops ?? 3));
    if (params.spaceId) url.searchParams.set("space_id", params.spaceId);
    if (params.direction) url.searchParams.set("direction", params.direction);
    if (params.relationTypes?.length) {
      url.searchParams.set("relation_types", params.relationTypes.join(","));
    }
    const res = await fetch(url.toString(), {
      method: "GET",
      headers: this.headers,
    });
    if (!res.ok) throw new RTMemoryError(res.status, await res.text());
    return res.json() as Promise<GraphNeighborhood>;
  }
}